Fixes #32437 - fix webpack-dev-server h2 issue

Ron-Lavi opened the pull request:
The issue today is that our old `webpack-dev-server` doesn't support the h2 protocol for node > v10
there is a workaround that all devs have to do in order to make it work, at least in the forklift's environment,
which is to go into `node_modules/webpack-dev-server/lib/Server.js` after each `npm install` and remove the `h2` protocol from the default list - which is exactly what I did in https://github.com/laviro/webpack-dev-server-without-h2
though it will make all our devs happier I assume.

The better solution would be to upgrade `webpack` and `webpack-dev-server`,
though this can take a while... sorry for this hack, I prefer having a node module that does it for us, instead of changing an existing node_module almost every day.

@theforeman/ui-ux I would appreciate your help on this one, thanks!

<!---

Thank you for contributing to The Foreman, please read the
[following guide](https://www.theforeman.org/contribute.html), in short:

* [Create an issue](https://projects.theforeman.org/projects/foreman/issues)
* Reference the issue via `Fixes #1234` in the commit message
* Prefer present-tense, imperative-style commit messages
* Mark all strings for translation, see [1]
* Suggest prerequisites for testing and testing scenarios following example above.
* Prepend `[WIP]` for work in progress to prevent bots from triggering actions
* Be patient, we will do our best to take a look as soon as we can
* Explain the purpose of the PR, attach screenshots if applicable
* List all prerequisites for testing (e.g. VMware cluster, two smart proxies...)
* Reviewers often use extensive list of items to check, have a look prior submitting [2]
* Be nice and respectful

1: https://projects.theforeman.org/projects/foreman/wiki/Translating#Translating-for-developers
2: https://github.com/theforeman/foreman/blob/develop/developer_docs/pr_review.asciidoc
-->

ShimShtein commented:
Another crazy idea:
create a npm/shell script that will patch the file directly and add the patch execution to the procfile.

Ron-Lavi commented:
> Another crazy idea:
> create a npm/shell script that will patch the file directly and add the patch execution to the procfile.

Could also work, but then I would suggest pinning the version of `webpack-dev-server` so we won't have any surprises.

tbrisker commented on config/webpack.config.js:
```diff
@@ -231,7 +231,8 @@ module.exports = env => {
       headers: { 'Access-Control-Allow-Origin': '*' },
-      hot: true
+      hot: true,
+      disableHostCheck: true,
```
We can override the same option here without requiring a forked version of the dev server
```suggestion
      disableHostCheck: true,
      https: { spdy: { protocols: ['http/1.1'] }},
```

Ron-Lavi replied:
I tried exactley to psss the same option, though it didn't work :(

ekohl replied:
I suspect passing `--https` as an argument may override the option here.

ShimShtein replied:
@tbrisker if I understand JS hash decomposition here: https://github.com/webpack/webpack-dev-server/blob/bc63a7191d947e030056c2c604743692c297b1b7/lib/Server.js#L537 the protocols node would be overridden, no matter what comes in options.https.

tbrisker replied:
@ShimShtein isn't that overriden only when `options.http2` is set? I'm guessing that @ekohl is right and that passing a command line argument might override the configuration

tbrisker replied:
I added a `console.log(options)` inside the dev server code right before the line @laviro changed and i did get my options there when running locally. when i add the `--https` indeed that option gets overriden though.

tbrisker replied:
Note that we are currently using version 2.11.5 of the webpack dev server, i wonder if we could upgrade the dev server without upgrading webpack itself?

tbrisker replied:
and here is where it gets overridden when `--https` is passed:
https://github.com/webpack/webpack-dev-server/blob/v2/lib/Server.js#L399-L410

ezr-ondrej replied:
I wonder if we could upgrade webpack :thinking:  xD

jeremylenz replied:
@MariaAga how's that going? :smile:

MariaAga replied:
@jeremylenz Slowly :laughing:

tbrisker commented:
[test foreman]
[test katello]

tbrisker reviewed (dismissed):
I'm not too happy about maintaining our own fork for this, but looks like webpack-dev-server version 2.x can't support https without http2 due to missing options when using command line parameters. Assuming tests pass now, approving this as a temporary workaround until we properly fix this issue and allow http2 connections.

MariaAga commented on script/foreman-start-dev:
```diff
@@ -1,3 +1,3 @@
 #!/bin/sh
-./node_modules/.bin/webpack-dev-server --config config/webpack.config.js --host "::" --disable-host-check $WEBPACK_OPTS &
+./node_modules/.bin/webpack-dev-server-without-h2 --config config/webpack.config.js --host "::" --disable-host-check $WEBPACK_OPTS &
```
Is `--disable-host-check` needed here if its specified `disableHostCheck: false` in the config?

Ron-Lavi replied:
thanks, removed it

MariaAga commented on config/webpack.config.js:
```diff
@@ -231,7 +231,8 @@ module.exports = env => {
       headers: { 'Access-Control-Allow-Origin': '*' },
-      hot: true
+      hot: true,
+      disableHostCheck: true,
```
@tbrisker I saw that the `disableHostCheck` option is not in the Procfile, is it ok that we're adding it?

tbrisker replied:
@laviro why is this change also required? I suppose it's ok since we don't necessarily have the host name configured properly in development and we don't really care about the hostnames in local development, but i might be missing something.

Ron-Lavi replied:
Sorry for not explaining more.. this is something I also encountered when spinning up a new forklift katello-devel machine, and was missing in the configuration for some reason

tbrisker replied:
@laviro can you please update the commit message explaining what change and why the changes are needed? see https://chris.beams.io/posts/git-commit/ and especially point 7 there.

ekohl replied:
@laviro IMHO that should be resolved, rather than worked around. Hosts should be correct and checked unless you can show that it really doesn't work. Usually it's a misconfiguration in your env.

ShimShtein replied:
@ekohl in my case, I have created local aliases in `/etc/hosts` to my machines, since the actual name is too long and autogenerated.

Ron-Lavi replied:
I also managed to fix it by adding the hostname to `config.hosts` via `config/environments/development.rb`
so I just removed it from our dev server config, although other devs might need to do the same step each time setting up a new environment.

MariaAga replied:
Can we document this somewhere? (In a different pr)

Ron-Lavi replied:
There's a warning shown in the UI that you need to add the hostname to `config.hosts`,
usually, it's set up for the `fqdn` from your settings, see: https://github.com/theforeman/foreman/blob/develop/config/environments/development.rb#L63
so I am not sure we need documentation for it, maybe the warning that's shown is enough

jeremylenz replied:
I just spun up a new `centos7-katello-devel-stable` and was unable to load any pages.  Adding my fqdn to `settings.yaml` didn't seem to do anything; it was only after adding back `disableHostCheck: true` here that I was able to get it working.

ronlavi2412 committed:
Fixes #32437 - Use webpack-dev-server-without-h2

Looks like webpack-dev-server version 2.x
can't support https without http2 due to missing options
when using command line parameters,
this is an hack that will prevent manually removing a line
after each `npm install` and will give us some air
while trying to upgrade to webpack 5.

Ron-Lavi commented:
@tbrisker how's that as a commit message? :)

Ron-Lavi commented:
ready for another review :)

MariaAga commented:
@laviro Looks good! Are the test failures related?

tbrisker commented:
[test foreman[

tbrisker commented:
[test foreman]

tbrisker reviewed (approved):
looks like webpack update is still not ready, lets get this in.
